applications/nrf_desktop/src/modules/dvfs.c

```diff
@@ -0,0 +1,260 @@
+ */
+static uint8_t dfvs_requests_state = BIT(DVFS_STATE_INITIALIZING);
+
+static enum dvfs_frequency_setting current_freq = DVFS_FREQ_HIGH;
+
+static struct dvfs_retry_work_data {
+	struct k_work_delayable dvfs_retry_work;
+	uint8_t dfvs_retries_cnt;
+} dvfs_retry_data;
+
+enum app_initial_task {
+	APP_INITIAL_TASK_SETTINGS_LOADER,
+	APP_INITIAL_TASK_BLE_BOND
+};
+
+static struct module_flags req_modules_bm;
+
+static const char *get_dvfs_frequency_setting_name(enum dvfs_frequency_setting setting)
+{
+	switch (setting) {
+	case DVFS_FREQ_HIGH: return "DVFS_FREQ_HIGH";
+	case DVFS_FREQ_MEDLOW: return "DVFS_FREQ_MEDLOW";
+	case DVFS_FREQ_LOW: return "DVFS_FREQ_LOW";
+	default: return "Unknown";
+	}
+}
+
+static const char *get_dvfs_state_name(enum dvfs_state state)
+{
+	switch (state) {
+	case DVFS_STATE_INITIALIZING: return "DVFS_STATE_INITIALIZING";
+	case DVFS_STATE_LLPM_CONNECTED: return "DVFS_STATE_LLPM_CONNECTED";
+	case DVFS_STATE_USB_CONNECTED: return "DVFS_STATE_USB_CONNECTED";
+	default: return "Unknown";
+	}
+}
+
+static void reschedule_dfvs_retry_work(enum dvfs_frequency_setting target_freq, bool is_initialized)
+{
+	if (dvfs_retry_data.dfvs_retries_cnt > DVFS_NUMBER_OF_RETRIES) {
+		LOG_ERR("DVFS retry count exceeded.");
+		module_set_state(MODULE_STATE_ERROR);
+		in_error_state = true;
+		return;
+	}
+	if (k_work_delayable_is_pending(&dvfs_retry_data.dvfs_retry_work)) {
+		return;
+	}
+	dvfs_retry_data.dfvs_retries_cnt++;
+	k_timeout_t timeout;
+
+	if (is_initialized) {
+		timeout = DVFS_RETRY_TIMEOUT;
+	} else {
+		timeout = DVFS_RETRY_INITIALIZATION_TIMEOUT;
+	}
+	(void) k_work_reschedule(&dvfs_retry_data.dvfs_retry_work, timeout);
+}
+
+static void cancel_dvfs_retry_work(void)
+{
+	(void) k_work_cancel_delayable(&dvfs_retry_data.dvfs_retry_work);
+	dvfs_retry_data.dfvs_retries_cnt = 0;
+}
+
+static void set_dvfs_freq(enum dvfs_frequency_setting target_freq)
+{
+	if (in_error_state) {
+		return;
+	}
+
+	size_t ret = dvfs_service_handler_change_freq_setting(target_freq);
+
+	if (ret == -EAGAIN) {
+		LOG_DBG("DVFS not initialized, trying again.");
+		reschedule_dfvs_retry_work(target_freq, false);
+
+	} else if (ret == -EBUSY) {
+		LOG_DBG("DVFS frequency change in progress");
+		reschedule_dfvs_retry_work(target_freq, true);
+
+	} else if (ret != 0) {
+		LOG_ERR("DVFS freq change returned with error: %d", ret);
+		module_set_state(MODULE_STATE_ERROR);
+	} else {
+		current_freq = target_freq;
+		LOG_INF("Have requested %s frequency",
+			get_dvfs_frequency_setting_name(target_freq));
+		cancel_dvfs_retry_work();
+	}
+}
+
+static enum dvfs_frequency_setting check_required_frequency(void)
+{
+	if (dfvs_requests_state & dvfs_high_freq_bitmask) {
+		return DVFS_FREQ_HIGH;
+	} else if (dfvs_requests_state & dvfs_medlow_freq_bitmask) {
+		return DVFS_FREQ_MEDLOW;
+	} else {
+		return DVFS_FREQ_LOW;
+	}
+}
+
+static void process_dvfs_states(enum dvfs_state state, bool turn_on)
```

Comment:
IMO opinion we should keep track of current DVFS preferred state. Lets imagine problem is caused by sysctrl and after a while we manage to recover from it - we might want to set DVFS in correct state and would need to know what frequency is needed.

Reply:
We do not have code to perform recovery after a fatal error so it may be reasonable to skip updates after we enter error state (we would not update DVFS frequency anyway)

Reply:
@mkapala-nordic what do you think?

Reply:
If we can't recover from fatal error then i do not see any reason why we should allow any operations then.

Reply:
I don't understand why we might not want to keep track of dfvs freq. for now we don't do it but in case of this module we might.  We do not stop other modules but maybe we should then on error stop the whole SoC as if the error happens it is probably sth with systemctrl and we should avoid working in such conditions imminently

Reply:
Discussed offline. Agreed to stop tracking DVFS state requests after fatal error (they would not be applied anyway).